### PR TITLE
panos new module used for security policy creation

### PR DESCRIPTION
**Pull request by ivanbojer**
##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
panos_security_policy

##### ANSIBLE VERSION

```
ansible 2.1.0.0 
config file = 
configured module search path = Default w/o overrides
```

##### SUMMARY
Security policies allow you to enforce rules and take action, and can be as general or specific as needed. The policy rules are compared against the incoming traffic in sequence, and because the first rule that matches the traffic is applied, the more specific rules must precede the more general ones.

### CODE REVIEW

**Comment by ivanbojer** [comment]
ready_for_review

**Comment by ivanbojer** [comment]
ready_for_review

**Comment by ivanbojer** [comment]
ready_for_review

File: lib/ansible/modules/network/panos/panos_security_policy.py [hunk 1]
```diff
@@ -0,0 +1,516 @@
...
+description: >
+    Security policies allow you to enforce rules and take action, and can be as general or specific as needed.
+    The policy rules are compared against the incoming traffic in sequence, and because the first rule that matches
+    the traffic is applied, the more specific rules must precede the more general ones.
+author: "Ivan Bojer (@ivanbojer)"
+version_added: "2.3"
+requirements:
+    - pan-python
+    - pandevice
+note:
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_security_policy.py]: `notes:` (plural)

File: lib/ansible/modules/network/panos/panos_security_policy.py [hunk 2]
```diff
@@ -0,0 +1,516 @@
...
+        default: None
+    devicegroup:
+        description: >
+            Device groups are used for the Panorama interaction with Firewall(s). The group must exists on Panorama.
+            If device group is not define we assume that we are contacting Firewall.
+        required: false
+        default: None
+    commit:
+        description:
+            - commit if changed
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_security_policy.py]: Commit if changed.

File: lib/ansible/modules/network/panos/panos_security_policy.py [hunk 3]
```diff
@@ -0,0 +1,516 @@
...
+        required: false
+        default: true
+'''
+
+EXAMPLES = '''
+# permit ssh to 1.1.1.1
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_security_policy.py]: Examples should specify a `- name:` rather than a comment, please see http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html#examples-block for details.

File: lib/ansible/modules/network/panos/panos_security_policy.py [hunk 4]
```diff
@@ -0,0 +1,516 @@
...
+    source_user: ['any']
+    destination: ['1.1.1.1']
+    category: ['any']
+    application: ['ssh']
+    service: ['application-default']
+    hip_profiles: ['any']
+    action: 'allow'
+    commit: false
+
+# Allow HTTP multimedia only from CDNs
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_security_policy.py]: Examples should specify a `- name:` rather than a comment, please see http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html#examples-block for details.

File: lib/ansible/modules/network/panos/panos_security_policy.py [hunk 5]
```diff
@@ -0,0 +1,516 @@
...
+    source_user: ['any']
+    destination: ['1.1.1.1']
+    category: ['content-delivery-networks']
+    application: ['http-video', 'http-audio']
+    service: ['service-http', 'service-https']
+    hip_profiles: ['any']
+    action: 'allow'
+    commit: false
+
+# more complex fictitious rule that uses profiles
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_security_policy.py]: Examples should specify a `- name:` rather than a comment, please see http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html#examples-block for details.

File: lib/ansible/modules/network/panos/panos_security_policy.py [hunk 6]
```diff
@@ -0,0 +1,516 @@
...
+    log_end: true
+    action: 'allow'
+    antivirus: 'default'
+    vulnerability: 'default'
+    spyware: 'default'
+    url_filtering: 'default'
+    wildfire_analysis: 'default'
+    commit: false
+
+# deny all
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_security_policy.py]: Examples should specify a `- name:` rather than a comment, please see http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html#examples-block for details.

File: lib/ansible/modules/network/panos/panos_security_policy.py [hunk 7]
```diff
@@ -0,0 +1,516 @@
...
+    username: 'admin'
+    password: 'paloalto'
+    rule_name: 'DenyAll'
+    log_start: true
+    log_end: true
+    action: 'deny'
+    rule_type: 'interzone'
+    commit: false
+
+# permit ssh to 1.1.1.1 using panorama and pushing the configuration to firewalls
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_security_policy.py:231]: Examples should specify a `- name:` rather than a comment, please see http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html#examples-block for details.

File: lib/ansible/modules/network/panos/panos_security_policy.py [hunk 8]
```diff
@@ -0,0 +1,516 @@
...
+    hip_profiles: ['any']
+    action: 'allow'
+    devicegroup: 'DeviceGroupA'
+'''
+
+RETURN = '''
+# Default return values
+'''
+
+ANSIBLE_METADATA = {'status': ['preview'],
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_security_policy.py]: Minor issue, not currently enforced by CI though may be soon, could you please move `ANSIBLE_METADATA` to before the DOCUMENTATION block. Thanks

The order is:
#!/usr/bin/perl
GPL
ANSIBLE_METADATA
DOCUMENTATION
EXAMPLES
RETURNS

File: lib/ansible/modules/network/panos/panos_security_policy.py [hunk 9]
```diff
@@ -0,0 +1,516 @@
...
+    from pan.xapi import PanXapiError
+    import pandevice
+    import pandevice.firewall
+    import pandevice.panorama
+    import pandevice.objects
+    import pandevice.policies
+
+    HAS_LIB = True
+except ImportError:
+    HAS_LIB = False
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_security_policy.py:270]: Where are you using `HAS_LIB`?

Generally after `argspec` you'd do something like
```
if not HAS_LIB:
    module.fail_json(msg="Missing requried pan-python module (check module docs)")
```

File: lib/ansible/modules/network/panos/panos_security_policy.py [hunk 10]
```diff
@@ -0,0 +1,516 @@
...
+    The policy rules are compared against the incoming traffic in sequence, and because the first rule that matches
+    the traffic is applied, the more specific rules must precede the more general ones.
+author: "Ivan Bojer (@ivanbojer)"
+version_added: "2.3"
+requirements:
+    - pan-python
+    - pandevice
+note:
+    - Checkmode is not supported.
+    - Panorama is supported
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_security_policy.py]: Include a URL detailing how to get the required libraries
` - pan-python can be obtained from U(http://example.com).`

File: lib/ansible/modules/network/panos/panos_security_policy.py [hunk 11]
```diff
@@ -0,0 +1,516 @@
...
+        required: false
+        default: true
+    action:
+        description:
+            - Action to apply once rules maches.
+        required: false
+        default: "allow"
+    group_profile:
+        description: >
+            security profile group that is already defined in the system. This property supersedes antivirus,
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_security_policy.py]: Security

File: lib/ansible/modules/network/panos/panos_security_policy.py [hunk 12]
```diff
@@ -0,0 +1,516 @@
...
+            - Username credentials to use for auth.
+        required: false
+        default: "admin"
+    password:
+        description:
+            - Password credentials to use for auth.
+        required: true
+    api_key:
+        description:
+            - API key that can be used instead of username/password credentials.
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_security_policy.py]: `API key that can be used instead of I(username)/I(password) credentials.`

http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html#formatting-options

**Review by gundalow — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/panos/panos_security_policy.py (hunk 1 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_security_policy.py]: OK

File: lib/ansible/modules/network/panos/panos_security_policy.py (hunk 10 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_security_policy.py]: OK, I added:
    - pan-python can be obtained from PyPi U(https://pypi.python.org/pypi/pan-python)
    - pandevice can be obtained from PyPi U(https://pypi.python.org/pypi/pandevice)

File: lib/ansible/modules/network/panos/panos_security_policy.py (hunk 11 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_security_policy.py]: OK

File: lib/ansible/modules/network/panos/panos_security_policy.py (hunk 12 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_security_policy.py]: OK

File: lib/ansible/modules/network/panos/panos_security_policy.py (hunk 2 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_security_policy.py]: OK

File: lib/ansible/modules/network/panos/panos_security_policy.py (hunk 3 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_security_policy.py]: OK

File: lib/ansible/modules/network/panos/panos_security_policy.py (hunk 4 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_security_policy.py]: OK

File: lib/ansible/modules/network/panos/panos_security_policy.py (hunk 5 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_security_policy.py]: OK

File: lib/ansible/modules/network/panos/panos_security_policy.py (hunk 6 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_security_policy.py]: OK

File: lib/ansible/modules/network/panos/panos_security_policy.py (hunk 7 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_security_policy.py:231]: OK

File: lib/ansible/modules/network/panos/panos_security_policy.py (hunk 8 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_security_policy.py]: OK, not a problem.

File: lib/ansible/modules/network/panos/panos_security_policy.py (hunk 9 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_security_policy.py:270]: Oversight and excellent catch! Done!

**Comment by ivanbojer** [comment]
@gundalow Thank you for the reviews. I've incorporated all the changes requested.

**Comment by ivanbojer** [comment]
ready_for_review

File: lib/ansible/modules/network/panos/panos_security_policy.py (hunk 8 above)
Comment (ivanbojer) [inline lib/ansible/modules/network/panos/panos_security_policy.py]: I tried this but then shippable run failed and I had to revert the changes.

Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_security_policy.py]: OK

**Comment by ivanbojer** [comment]
ready_for_review

**Comment by ivanbojer** [comment]
bot_status

**Comment by ivanbojer** [comment]
bot_broken

**Comment by mattclay** [comment]
Closing and re-opening to trigger CI.

**Comment by ivanbojer** [comment]
ready_for_review

File: lib/ansible/modules/network/panos/panos_security_policy.py [hunk 13]
```diff
@@ -0,0 +1,518 @@
...
+        wildfire_analysis=dict(),
+        log_start=dict(type='bool', default=False),
+        log_end=dict(type='bool', default=True),
+        rule_type=dict(default='universal'),
+        action=dict(default='allow'),
+        devicegroup=dict(),
+        commit=dict(type='bool', default=True)
+    )
+    module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=False,
+                           required_one_of=[['api_key', 'password']])
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_security_policy.py:401]: Do you also need multulary_exclusive as well?

File: lib/ansible/modules/network/panos/panos_security_policy.py [hunk 14]
```diff
@@ -0,0 +1,518 @@
...
+
+        changed = add_security_rule(device, sec_rule)
+    except PanXapiError:
+        exc = get_exception()
+        module.fail_json(msg=exc.message)
+
+    if changed and commit:
+        result = _commit(device, devicegroup)
+
+    module.exit_json(changed=changed, msg="okey dokey")
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_security_policy.py:489]: Is there a better `msg` we can return?

File: lib/ansible/modules/network/panos/panos_security_policy.py [hunk 15]
```diff
@@ -0,0 +1,518 @@
...
+        default: None
+    devicegroup:
+        description: >
+            Device groups are used for the Panorama interaction with Firewall(s). The group must exists on Panorama.
+            If device group is not define we assume that we are contacting Firewall.
+        required: false
+        default: None
+    commit:
+        description:
+            - Commit if changed
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_security_policy.py]: Missing full stop. Is their any more detail we can add? Does panos support commit messages, if so what is used?

File: lib/ansible/modules/network/panos/panos_security_policy.py [hunk 16]
```diff
@@ -0,0 +1,518 @@
...
+        required: true
+    username:
+        description:
+            - Username credentials to use for auth.
+        required: false
+        default: "admin"
+    password:
+        description:
+            - Password credentials to use for auth.
+        required: true
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_security_policy.py]: Not required if one of password or api_key is required. 

Descriptions for password and api_key should both state. 
"Exactly one of I(password) or I(api_key) is required."

File: lib/ansible/modules/network/panos/panos_security_policy.py [hunk 17]
```diff
@@ -0,0 +1,518 @@
...
+    api_key:
+        description:
+            - API key that can be used instead of I(username)/I(password) credentials.
+    rule_name:
+        description:
+            - Name of the security rule.
+        required: true
+    rule_type:
+        description:
+            - Type of security rule (6.1+).
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_security_policy.py]: Should this say something like:
Only supported on PanOS 6.1+ devices

File: lib/ansible/modules/network/panos/panos_security_policy.py [hunk 18]
```diff
@@ -0,0 +1,518 @@
...
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+ANSIBLE_METADATA = {'status': ['preview'],
+                    'supported_by': 'community',
+                    'version': '1.0'}
+
+DOCUMENTATION = '''
+---
+module: panos_security_policy
+short_description: create security rule policy
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_security_policy.py]: Short_descriptions need to be meaningful without knowing what category (directory structure) the module is is,  how about something like:
Create security rule policy on PanOS  devices

**Review by gundalow — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/panos/panos_security_policy.py (hunk 18 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_security_policy.py]: OK I will change that. Given that I work with PanOS devices mainly I guess I am seeing customer environments with only PanOS devices.

File: lib/ansible/modules/network/panos/panos_security_policy.py (hunk 16 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_security_policy.py]: It is a bit more complicated than that as it is either username/password or API key. The code is smart enough to figure out internally whether user sets API key or username/password.

I went with 'Username credentials to use for auth unless I(api_key) is set.'

File: lib/ansible/modules/network/panos/panos_security_policy.py (hunk 17 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_security_policy.py]: OK

File: lib/ansible/modules/network/panos/panos_security_policy.py (hunk 15 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_security_policy.py]: I added full stop. PanOS has a concept of candidate configuration that once committed will become running configuration.

File: lib/ansible/modules/network/panos/panos_security_policy.py (hunk 13 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_security_policy.py:401]: It is not really mutually exclusive as code logic will try multiple authentication methods. It is more of 'unnecessary' case.

File: lib/ansible/modules/network/panos/panos_security_policy.py (hunk 14 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_security_policy.py:489]: We could return the result but it is probably too much for average Ansible user. I will stick with our typical error message we used before.

**Comment by ivanbojer** [comment]
bot_status

**Comment by ivanbojer** [comment]
It seems the same issue like above where automatic build has not been triggered.

**Comment by ivanbojer** [comment]
@gundalow thank you for review!

ready_for_review

**Comment by mattclay** [comment]
With this issue flagged as broken for the bot, it doesn't get picked up for review for CI issues. The bot isn't responsible for CI, so having it flagged as broken for CI related issues is counter productive.

**Comment by mattclay** [comment]
Closing and re-opening PR to trigger CI.

**Comment by mattclay** [comment]
I've removed the comments telling the bot it is broken, as well as the label. That should restore the PR to the pre-broken state and allow the bot to function normally.

**Comment by mattclay** [comment]
Looks like the bot is functioning properly now.

**Comment by ivanbojer** [comment]
@mattclay Thank you for fixing it.

**Comment by ivanbojer** [comment]
ready_for_review

**Comment by ivanbojer** [comment]
@gundalow All changes are incorporated, any comments / ready to go?

**Review by gundalow — APPROVED** [review]
